Handle Variables in the URL

AnnMarieW opened the pull request:
Closes #45 

This is a new feature for handing variables in the URL.

Creates a new parameter `path_template` in `dash.page_registry` 

- `path_template`:
       Add variables to a URL by marking sections with <variable_name>. The layout function
       then receives the <variable_name> as a keyword argument.
       e.g. path_template= "/asset/<asset_id>"
             path = "/assets/a100" 
            layout will receive `{"asset_id":"a100"}`

If no `path` is supplied when calling `dash.register_page()` then it will be inferred from the `path_template`.  Variables will default to `"none"`.
  e.g. "/asset/<asset_id>"  => "/asset/none"

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -259,12 +277,20 @@ def router():
+                    template_id = app.strip_relative_path(page["path_template"])
+                    path_variables = _parse_path_variables(path_id, template_id)
+                    if path_variables:
+                        layout = page["layout"]
```
🐎 do we want to `break` here and [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR286) since we found a matching page?

AnnMarieW replied:
For sure - that would be more efficient

AnnMarieW replied:
actually, I could add a `break` here but not [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR286) because above can be true for paths with templates

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -398,3 +430,24 @@ def _parse_query_string(search):
+        if template_segments[i].startswith("<"):
+            # removes the "<" ">" around the variable key
+            variable_key = template_segments[i][1:-1]
+            path_segments[i] = None if path_segments[i] == "None" else path_segments[i]
```
This `"None"` -> `None` conversion worries me - what if you really did want to pass `"None"` as the parameter? Can we just leave it as `"None"`? I know up [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR247) we use `"None"` in the path when registering the page but I'm not sure this is a good idea either - for example, how is that going to work in the title, description, and image?

https://github.com/plotly/dash-labs/blob/f182bc011291fedce47c13e6a889d09b17a7c020/dash_labs/plugins/pages.py#L324-L332

I suspect the better way to handle this will be to take the [path-to-layout finder](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR282-R292) and turn it into a path-to-page finder, then use that both here and in `interpolate_index`

alexcjohnson replied:
In fact, I could imagine users wanting title, description, and image to each potentially be a function of the path variables... But for now we should just ensure they work correctly as constant values for each registered page.

AnnMarieW replied:
Actually this works without any changes.  I updated the example app to include `title` and `description`.  Here's what the meta tags look like when no initial `path` is provided:
 
![image](https://user-images.githubusercontent.com/72614349/145596467-0435fa13-2311-4b97-82fa-3ef9825076d1.png)

And here is what it looks like when a different (valid) initial path is provided:
![image](https://user-images.githubusercontent.com/72614349/145596883-cc920c4d-a326-4775-941e-44952c7352ee.png)

AnnMarieW replied:
> In fact, I could imagine users wanting title, description, and image to each potentially be a function of the path variables... But for now we should just ensure they work correctly as constant values for each registered page.

I think one way to do this is to create a `dash.path_registry`.  It could start by being generated from `dash.page_registry`  Key would be the `path`, values would include `title, description, image`.  Then people could add other paths.

AnnMarieW replied:
>I suspect the better way to handle this will be to take the path-to-layout finder and turn it into a path-to-page finder, then use that both here and in interpolate_index.

Good suggestion - I added a `path_to_page` function in  [this commit](https://github.com/plotly/dash-labs/pull/61/commits/2ddad2ca6f89c45e35fc05fd90139d5cd4276c96)

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -398,3 +430,24 @@ def _parse_query_string(search):
+        return None
+
+    path_vars = {}
+    for i in range(len(path_segments)):
```
this would be a little cleaner as:
`for path_segment, template_segment in zip(path_segments, template_segments):`
since you already know the lengths are the same.

That said I suspect at some point we'll want to extend this to support [Flask converter types](https://flask.palletsprojects.com/en/2.0.x/quickstart/#variable-rules) including `path` that also allows slashes in the match, but that will need some special logic anyway.

AnnMarieW replied:
I thought about doing the zip, but then there would be additional logic to pull out the variables

alexcjohnson replied:
Where? All I see is `path_segments[i]` which would just turn into `path_segment`, and similarly with `template_segments[i]`->`template_segment`. You never use `i` for any purpose other than pulling out the ith element of those two lists, so you should be able to get rid of it entirely.

AnnMarieW replied:
I have a new version [here](https://github.com/plotly/dash-labs/pull/61/commits/9c5e92bc6cd5c9c4437b1a64face96ea085bc5a2) is it better?

I also removed the conversion from "None" to `None`.   

Also, When no `path` is supplied to  `dash.page_registry` I changed the path variables in the inferred path  "none"  (instead of "None").  I hope this makes it more clear that it's a string and not actually `None`

AnnMarieW commented:
Hey @alexcjohnson This is ready for another review.  I moved tests to #66 so it doesn't hold up things that are finished, such as:


- Fixed the "None" => `None` conversion.
- Confirmed that when there are variables in the path, the meta tags for title, description, and image are correct when the page first loads - both when the path is supplied or inferred.  However it doesn't update when the path variables change.  If that's a popular feature, then I think it could be added later.
- The `assets/` and `pages/` folders are now optional.  If  `pages/` can't be found,  it no longer raises  an Exception,  but it does print a message.  The message  could help debug if not finding  `pages/` is unintentional.


- There is now a check that the paths are unique for each module in dash.page_registry. This should help prevent bugs.
- There is now a check that pathnames start with a `/`.   Previously the `/` was missing in some cases when the layout was added directly to dash.register_page()

- Updated docs and included new info on path variables and query strings

chriddyp commented on docs/08-MultiPageDashApp.md:
```diff
@@ -285,47 +323,63 @@ page navigation links or by template authors.
    e.g. `pages.weekly_analytics` to `/weekly-analytics`
 
+
+- path_template:
```
```suggestion
- `path_template`:
```

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -9,22 +9,28 @@
 _ID_LOCATION = "_pages_plugin_location"
-_ID_DUMMY = "_pages_plugin_dummy"
+_ID_STORE = "_pages_plugin_store"
+_ID_DUMMY = "pages_plugin_dummy"
```
```suggestion
_ID_DUMMY = "_pages_plugin_dummy"
```
Was that change intentional?

AnnMarieW replied:
oops that was a typo

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -224,16 +241,17 @@ def _filename_to_name(filename):
+        return path
+    else:
+        # replace the variables in the template with "None"
+        return re.sub("<.+?>", "none", template)
```
```suggestion
        return re.sub("<[^<>]+>", "none", template)
```
Not a big deal, I'm just brainstorming ways the user could break this, and at present unmatched or empty angle brackets would produce very weird results.

Maybe in fact the best way to handle this would be validating the template in `register_page`: break it into segments like we do in `_parse_path_variables`, but then for any segment that has either `<` or `>` in the segment, throw an error if the segment doesn't match `^<[^<>]+>$`

AnnMarieW replied:
ah, yes - good idea. I'll add that

alexcjohnson reviewed (approved):
Looks good! Stricter pattern checking https://github.com/plotly/dash-labs/pull/61/files#r778960479 would be nice but we could do that later as well.
💃

alexcjohnson commented on dash_labs/plugins/pages.py:
```diff
@@ -224,16 +243,27 @@ def _filename_to_name(filename):
+def _validate_template(template):
+    template_segments = template.split("/")
+    for s in template_segments:
+        if ("<" or ">") in s and not (s.startswith("<") and s.endswith(">")):
```
`("<" or ">") in s`  -> I think you mean `("<" in s or ">" in s)` 

Also need to check that `<` and `>` are NOT in `s[1: -1]`

AnnMarieW replied:
> Also need to check that < and > are NOT in s[1: -1]

Just to summarize our discussion:  Since the variables are passed to the layout in a dict as strings, there should not be a restriction about including  `<` or ` >` .   However, the most common use case is that the keys are Python variables, so now there is a warning message if the variable is not a valid Python variable name.

alexcjohnson reviewed (approved):
Alright I think we have a winner! 💃
